File: problem45.py
import List
class Solution:
    def jump(self, nums: List[int]):
        # Base case
        if len(nums) == 1:
            return 0
        steps = 0
        farthest = 0 # farthest jump from current index
        end = 0 # end of current jump
        # Save concept
        # Greedy to reach final distition with least steps
        for i in range(len(nums)-1):
            farthest = max( farthest, i + nums[i])

            # we will increase the number of step, we reach the end of current possible step
            # then we reach the end, we know the farthest step for next step.
            if i == end:
                steps += 1
                end = farthest
        return steps

File: problem45.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    int jump(vector<int> &nums)
    {
        int steps = 0;   // track the lowest step
        int farjump = 0; // farthest i can jump the current index
        int end = 0;     // this the end farthest at current step

        for (int i = 0; i < nums.size() - 1; i++)
        {
            farjump = max(farjump, i + nums[i]);

            if (end == i)
            {
                steps++;
                end = farjump;
            }
        }
        return steps;
    }
};